MovingBlocks/Terasology
# feat: set limits for maximum memory use.

We've had some memory leaks where the game seems to take up far more of the system's memory than the amount allowed for Java's maximum heap size.

This uses some operating-system-specific functions to set limits on how much memory the process can allocate.

Because this is so system-dependent, I'm introducing this as a developer/debugger feature rather than intending it for production use.

(Players won't want the game taking runaway amounts of memory either, but getting consistent support for this and setting expectations for how it works would be a bigger project. Possibly would want to do it in Launcher, as some of these controls are better set before even launching the process.)

### Contains

Linux support only.

* Sets [RLIMIT_DATA](https://www.man7.org/linux/man-pages/man3/getrlimit.3p.html#DESCRIPTION) to limit the maximum amount of memory for the process's data segment.
* Sets [`oom_score_adj`](https://www.kernel.org/doc/html/latest/filesystems/proc.html#proc-pid-oom-adj-proc-pid-oom-score-adj-adjust-the-oom-killer-score) to give the out-of-memory reaper encouragement to choose this process for termination when the system is out of memory as a whole.

### How to test

On Linux:

check `/proc/PID/limits`

* test default
* test with low limit, see out-of-memory error
* test with high limit

On other platforms:

* make sure game starts as normal, i.e. that the Linux-only code didn't ruin things for everyone else.


### Outstanding before merging

- [x] Make limits configurable instead of hard-coded
- [x] Blocked by #4146 for the upgrade of JNA Platform
